doomworld_downloader/doomworld_downloader.py

```diff
@@ -94,55 +127,116 @@ def parse_thread_page(base_url, page_number):
-            else:
-                last_page_num = 1
-
-        for page_num in range(last_page_num, 0, -1):
-            cur_posts = parse_thread_page(thread.url, page_num)
-            new_posts = [post for post in cur_posts if post.post_date > search_start_date and post.post_date < search_end_date]
+        # In case testing, uncomment to speed up the tests
+        # break
+        for page_num in range(thread.last_page_num, 0, -1):
+            cur_posts = parse_thread_page(thread.url, page_num, thread)
+            # If the last post on a page is before the start date, we can break out immediately
+            # since we are going backwards in time from the last page.
+            if cur_posts and cur_posts[-1].post_date < search_start_date:
+                break
+
+            new_posts = [post for post in cur_posts
+                         if search_start_date < post.post_date < search_end_date]
             posts.extend(new_posts)
-            if len(cur_threads) != len(new_threads):
+            if not new_posts:
                 break
 
-    print(posts)
+    # In case testing, uncomment in case you need just a couple test posts
+    # posts = [Post(author_name='the_kovic', post_date=datetime(2020, 4, 10, 15, 22, 55), attachments={'kovic_e2m1-40.zip': 'https://www.doomworld.com/applications/core/interface/file/attachment.php?id=82293'}, links={'https://www.youtube.com/watch?v=aXyPH0J4BD8': 'https://www.youtube.com/watch?v=aXyPH0J4BD8'}, post_text='Ultimate Doom E2M1 in 40\nPort used: Crispy Doom\nDemo:\nVideo:\nI was inspired to attempt to run something in Doom by a couple of content creators on YT (you probably know which ones), decided to try E2M1. I got 41 in about ten minutes and then spent three more hours grinding 40. It might be bias but for now I think that running Doom is much harder for me than bunnyhopping in Source games (which is what I usually play and run).\nI hope I read all the rules correctly and that the demo works fine.', parent=Thread(name='Personal Best Demo Thread ← POST YOUR NON-WRs HERE', id=112532, url='https://www.doomworld.com/forum/topic/112532-personal-best-demo-thread-%E2%86%90-post-your-non-wrs-here/', last_post_date=datetime(2020, 4, 11, 3, 4, 56), last_page_num=3)), Post(author_name='RobUrHP420', post_date=datetime(2020, 4, 11, 3, 4, 56), attachments={'9.94e1m1(uv  pacifist).zip': 'https://www.doomworld.com/applications/core/interface/file/attachment.php?id=82370'}, links={'https://www.youtube.com/watch?v=2LF_jlA1aLc': 'https://www.youtube.com/watch?v=2LF_jlA1aLc'}, post_text='Finally hit 9s on Hangar (UV Pacifist) So happy rn! Took me well over a thousand attempts.\nVideo:', parent=Thread(name='Personal Best Demo Thread ← POST YOUR NON-WRs HERE', id=112532, url='https://www.doomworld.com/forum/topic/112532-personal-best-demo-thread-%E2%86%90-post-your-non-wrs-here/', last_post_date=datetime(2020, 4, 11, 3, 4, 56), last_page_num=3))]
+    # print(posts)
+
+    demo_jsons = []
+    for post in posts:
```

Comment:
Maybe an early step we can make here is to just map out the skeleton of what kind of objects we need and get things into compartmentalized files where people can edit the separate pieces in parallel (we hope anyway :wink: )

Something that takes a block of text and guesses demo details, something that gets file data from the lmp, something that runs the lmp through prboom+, something that converts a levelstat file into useful data, etc.

And most likely those things should operate on a special object that's mergeable so that you can collect a bunch of chunks and put it all together. It could even track disputes.

Something like `{ name: { certain: "4shockblast" }, category: { possible: ["Pacifist", "UV Speed"] }`

Just thinking out loud :smile: